scipy/_lib/_array_api.py

```diff
@@ -571,3 +584,39 @@ def xp_vector_norm(x: Array, /, *,
     else:
         # to maintain backwards compatibility
         return np.linalg.norm(x, ord=ord, axis=axis, keepdims=keepdims)
+
+# utility to broadcast arrays and promote to common dtype
+def xp_broadcast_promote(*args, ensure_writeable=False, xp=None):
```

Comment:
Yes, `None` was a pain. Typically there is one `None` at most, and usually it is the last argument. Handling them more generally was a bit of overkill to avoid adding caveats to the documentation.

I would prefer for the functions that call this not to need to branch for `None`s. That's why I put in the effort for this to handle `None`s. I'm sure glad `array_namespace` handles `None`.

That said, if this needs to be simplified, not handling `None`s would simplify it quite a bit.

Reply:
I think that it is good to have this helper scale well for a greater quantity of `*args` than 2, and it would quickly become more ugly to have to handle `None`s at the point of use for 4 or 5 or so arrays. While it isn't clean to handle `None`s here, I don't think that really matters given that this is hidden away in `_lib`. And it isn't *too* bad, so I'm happy to keep the `None` handling here. We could always remove it in the future.